### ISSUE TITLE
Experiment: //cmd/ko

### ISSUE THREAD

**Issue by mattmoor** [google/go-containerregistry#80]
**tl;dr** I want a more purpose built variation of what you get with `bazel` + `rules_go` + `rules_docker` + `rules_k8s` that simply wraps the Go toolchain directly.

### Background: State of Bazel

Today in Bazel, my build is described in `BUILD` files (largely generated by Gazelle) and extended to handle containerization and kubernetesification via:
```python
# BEGIN generated by Gazelle
go_library(
    name = "go_default_library",
    ...
)

go_test(
    name = "go_default_test",
    ...
)

go_binary(
    name = "name-of-directory",
    ...
)
# END generated by Gazelle

# Wrap the Go binary into a minimal container image.
go_image(
    name = "image",
    binary = ":name-of-directory",
)

# Helper for interacting with my K8s "Deployment"
k8s_object(
    name = "deployment",
    template = "deployment.yaml",
    images = {
        # Associate the image reference in deployment.yaml with the associated binary target.
        "gcr.io/foo/bar:baz": ":image",
    },
)
```

With this, I can quickly iterate and redeploy via `bazel run :deployment.apply`.

Unfortunately, it's painful to deal with Go in Bazel because it forks the source of truth for dependencies: the import paths in `*.go`.  This often leaves you running Gazelle before build commands, which breaks the promise of `rules_k8s`: a single command for rapid iteration.

You also end up checking in a lot of redundant configuration.

### Background: Go importpaths

As alluded to above, Go declares dependencies via a block like this at the top of each file:
```go
import (
    "github.com/google/go-containerregistry/authn"
    "github.com/google/go-containerregistry/name"
)
```

Binary targets are determined by the declaration of `package main` within a particular one of these paths.  For example, we have a binary that can be referenced via the Go importpath: `github.com/google/go-containerregistry/cmd/d8s`

#### Foreshadowing

Gee, Go importpaths look a *LOT* like Docker image references.

### Convention over Configuration

This idea builds around an extension of the Go importpath convention to tie Go binaries all the way through to one's Kubernetes configuration.  Instead of writing an image reference in your Kubernetes configuration, you would write a Go importpath:
```yaml
apiVersion: apps/v1beta1
kind: Deployment
metadata:
  name: foo
spec:
  replicas: 1
  template:
    spec:
      containers:
      - name: bar
        image: github.com/path/to/my/thing
```

In Bazel, we had to manually associate our `BUILD` target for `go_image` with the image reference in the template.  In this example, we would implicitly associate embedded Go importpath references to a binary, which we'd wrap in a container in a similar fashion to `go_image`.

This set of conventions should enable us to have something like `rules_k8s` that's driven purely by conventions without forking any sources of truth.

#### Hold your horses

... unless Github is launching a Docker registry, this is still a fantasy.  We can't publish images to those embedded references!

Luckily we've seen and solved this same problem with Bazel already, via: `image_chroot`.  The value of `image_chroot` is used to prefix image references within the templates (for things we plan to publish).  So `image_chroot = "gcr.io/foo"` with an `image: github.com/bar/baz` would become the image reference: `image: gcr.io/foo/github.com/bar/baz`.  This *looks* funny, but it is a legitimate Docker image URI.

*Unfortunately, this will not work with certain registries (including DockerHub), which disallow multi-level image names.  I'm calling this an acceptable loss and not going to lose a lot of sleep over it, at least for now.*

### Ok, Some Configuration...

So we will still have at least a little configuration for the typical flow (to configure where to publish images), but there are also some pretty obvious knobs that folks will look for pretty quickly (e.g. what base image?).

I think for the purposes of prototyping, I'm going to just use `$KO_DOCKER_REPO` for the base repository with which to chroot image references.

Over time, I think the most idiomatic solution for configuration will be to adopt something like `Kopkg.toml` (and `Kopkg.lock`).  This would enable us to reference a base image (tag) and lock it (digest) for reproducible builds.

#### Extra Credit

Given the incredible portability of Go, I can easily see us extending this to produce "fat" images (aka "manifest lists") pretty trivially.  I think that this should motivate us to structure our configuration in a certain way:
```toml
# TODO(mattmoor): We want this to be something individual developers can
# override (e.g. stamp variables), so perhaps KO_DOCKER_REPO could still override this?
repository = "gcr.io/my-project"

[[linux/amd64]]
  base = "gcr.io/base-images/blah:amd64"

[[linux/arm]]
  base = "gcr.io/base-images/blah:arm"

[[windows/amd64]]
  base = "gcr.io/base-images/blah:win64"
```

### The `ko` CLI Surface

The goal of `ko` is to effectively blend:
1. The CLI surface of `go` (e.g. `build`, `test`, ...)
1. An enhanced subset of `kubectl` (e.g. `apply`, `create`, `update`, `delete`)

The "enhancement" of the latter is that we will build, publish, and resolve image references within templates in a similar manner to `rules_k8s`.

#### Tying it all together

With this surface, I should be able to have a fast-path development experience similar to `bazel run :deployment.apply` by simply running: `ko apply -f deployment.yaml`

**Comment by mattmoor** [comment]
I forgot to mention something notable above...  Just as a `bazel build`-able repo can also be `go build`-able (e.g. the many K8s repos), the `ko` conventions will work with `rules_k8s` so you should be able to start using `ko` with repos currently using `rules_k8s`, if you adopt the `ko` convention.

**Comment by mattmoor** [comment]
re: configuration

We will want to support different base images for different packages.  e.g. `github.com/foo/bar` may need something different from `github.com/baz/blah`.

Still mulling how we want to express this.  I think we want a "use this everywhere" and a "use this here" that trumps it.  Ideas welcome.
